common/meta/grpc/src/grpc_client.rs

```diff
@@ -95,20 +96,61 @@ impl MetaChannelManager {
     }
 }
 
+struct RoundRobinAddr {
+    addrs: Vec<String>,
+    idx: Arc<RwLock<usize>>,
+}
+
+impl RoundRobinAddr {
+    fn new(addrs: Vec<String>) -> Self {
+        Self {
+            addrs,
+            idx: Arc::new(RwLock::new(rand::thread_rng().gen_range(0..10))),
+        }
+    }
+
+    async fn next(&self) -> Option<String> {
+        let mut idx = self.idx.write().await;
+        *idx = (*idx + 1) % self.addrs.len();
+        let addr = self.addrs.get(*idx);
+        addr.map(|a| a.to_string())
+    }
+}
+
 #[async_trait]
 impl ItemManager for MetaChannelManager {
     type Key = Vec<String>;
     type Item = Channel;
     type Error = MetaError;
 
     async fn build(&self, endpoints: &Self::Key) -> std::result::Result<Self::Item, Self::Error> {
-        let channel_eps: std::result::Result<Vec<Endpoint>, MetaError> = (*endpoints)
-            .iter()
-            .map(|a| self.build_endpoint(a))
-            .collect();
-        let channel_eps = channel_eps?;
-        let ch = Channel::balance_list(channel_eps.into_iter());
-        Ok(ch)
+        let mut n = endpoints.len();
+        let addrs = RoundRobinAddr::new(endpoints.clone());
+        let item = loop {
+            if let Some(addr) = addrs.next().await {
```

Comment:
if all the addr is invalid, it may cause a dead loop?

Reply:
Right, a for-loop looks good enough

Reply:
@lichuang if all addrs are invalid, will break loop at [here](https://github.com/datafuselabs/databend/pull/5616/files#diff-757d3d716b4424b8e86d968f8a68c94b6bc9cd938ac9ee3caa9440ef35bb7707R143) with last error.